// mangoose-migrate/dist/operations/index.js
// src/operations/create-model.ts
import mongoose from "mongoose";
var CreateModel = class {
  constructor(modelName, schemaDefinition) {
    this.modelName = modelName;
    this.schemaDefinition = schemaDefinition;
  }
  async applyUp(connection) {
    const schema = new mongoose.Schema(this.schemaDefinition);
    connection.model(this.modelName, schema);
  }
  async applyDown(connection) {
    connection.deleteModel(this.modelName);
  }
};

// src/operations/add-field.ts
var AddField = class {
  constructor(modelName, fieldName, fieldDefinition) {
    this.modelName = modelName;
    this.fieldName = fieldName;
    this.fieldDefinition = fieldDefinition;
  }
  async applyUp(connection) {
    const model = connection.model(this.modelName);
    const schema = model.schema;
    schema.add({ [this.fieldName]: this.fieldDefinition });
    connection.model(this.modelName, schema);
  }
  async applyDown(connection) {
    const model = connection.model(this.modelName);
    const schema = model.schema;
    schema.remove(this.fieldName);
    connection.model(this.modelName, schema);
  }
};
export {
  AddField,
  CreateModel
};
